classes/local/secret_manager.php

```diff
@@ -0,0 +1,228 @@
+     * @return void
+     */
+    private function add_secret_to_db(string $secret, int $expires, string $sessionid = null) {
+        global $DB, $USER;
+        $expirytime = time() + $expires;
+
+        $data = [
+            'userid' => $USER->id,
+            'factor' => $this->factor,
+            'secret' => $secret,
+            'timecreated' => time(),
+            'expiry' => $expirytime,
+            'revoked' => 0
+        ];
+        if (!empty($sessionid)) {
+            $data['sessionid'] = $sessionid;
+        }
+        $DB->insert_record('tool_mfa_secrets', $data);
+    }
+
+    /**
+     * Validates whether the provided secret is currently valid.
+     *
+     * @param string $secret the secret to check
+     * @return string a secret manager state constant
+     */
+    public function validate_secret(string $secret) : string {
+        global $DB, $USER;
+        $status = $this->check_secret_against_db($secret, $this->sessionid);
+        if ($status !== self::NONVALID) {
+            if ($status === self::VALID) {
+                // Cleanup DB $record.
+                $DB->delete_records('tool_mfa_secrets', ['userid' => $USER->id, 'factor' => $this->factor]);
+            }
+            return $status;
+        }
+        // This is always nonvalid.
+        return $status;
+    }
+
+    /**
+     * Checks if a given secret is valid from the Database.
+     *
+     * @param string $secret the secret to check.
+     * @param string $sessionid the session id to check for.
+     * @return string a secret manager state constant.
+     */
+    private function check_secret_against_db(string $secret, string $sessionid) : string {
+        global $DB, $USER;
+
+        $sql = "SELECT *
+                  FROM {tool_mfa_secrets}
+                 WHERE secret = :secret
+                   AND expiry > :now
+                   AND userid = :userid
+                   AND factor = :factor";
+
+        $params = [
+            'secret' => $secret,
+            'now' => time(),
+            'userid' => $USER->id,
+            'factor' => $this->factor
+        ];
+
+        $record = $DB->get_record_sql($sql, $params);
+
+        if (!empty($record)) {
+            // If revoked it should always be revoked status.
+            if ($record->revoked) {
+                return self::REVOKED;
+            }
+
+            // Check if this is valid in only one session.
+            if (!empty($record->sessionid)) {
+                if ($record->sessionid === $sessionid) {
+                    return self::VALID;
+                }
+                return self::NONVALID;
+            }
+            return self::VALID;
+        }
+        return self::NONVALID;
+    }
+
+    /**
+     * Revokes the provided secret code for the user.
+     *
+     * @param string $secret the secret to revoke.
+     * @return void
+     */
+    public function revoke_secret(string $secret) {
```

Comment:
I think this should have $user passed in not $USER, if you revoke from a url in a text or email you might not be logged in yourself (yet) and I don't think we should force people to login then, as they might not be able to if their account is already compromised

Reply:
Added $userid as an optional param, defaulting to $USER if null. Makes it easier to use for inband operations, while providing scope for out of band